File: src/hooks.rs
use crate::{
    fsio::{goal_integrity, read_policy},
    models::HookInput,
};
use anyhow::{Context, Result};
use serde_json::json;
use std::{
    io::{self, Read},
    path::{Path, PathBuf},
};

pub(crate) fn pretooluse(root: &Path) -> Result<()> {
    let mut raw = String::new();
    io::stdin()
        .read_to_string(&mut raw)
        .context("read hook stdin")?;
    if raw.trim().is_empty() {
        return approve();
    }

    let input: HookInput = serde_json::from_str(&raw).context("parse hook input")?;
    let policy = read_policy(root)?;
    if !policy.active {
        return approve();
    }

    if is_mutating_tool(&input.tool_name) {
        if let Err(err) = goal_integrity(root) {
            return deny(format!("GOAL.html changed while loop is active: {err:#}"));
        }
    }

    if input.tool_name.eq_ignore_ascii_case("bash") {
        if let Some(command) = input
            .tool_input
            .get("command")
            .and_then(|value| value.as_str())
        {
            let lower_command = command.to_ascii_lowercase();
            for denied in &policy.deny_shell_substrings {
                if lower_command.contains(&denied.to_ascii_lowercase()) {
                    return deny(format!("command contains denied substring {denied:?}"));
                }
            }
        }
    }

    for path in hook_file_paths(root, &input) {
        let normalized = normalize_path_for_policy(root, &path);
        if path_matches_any(&normalized, &policy.immutable_paths) {
            return deny(format!("{normalized} is immutable while yoloop is active"));
        }
        if path_matches_any(&normalized, &policy.protected_paths_while_active) {
            return deny(format!("{normalized} is protected while yoloop is active"));
        }
    }

    approve()
}

fn approve() -> Result<()> {
    println!("{}", json!({ "decision": "approve" }));
    Ok(())
}

fn deny(reason: String) -> Result<()> {
    println!("{}", json!({ "decision": "block", "reason": reason }));
    Ok(())
}

fn is_mutating_tool(tool_name: &str) -> bool {
    matches!(
        tool_name.to_ascii_lowercase().as_str(),
        "bash" | "write" | "edit" | "multiedit" | "notebookedit"
    )
}

fn hook_file_paths(root: &Path, input: &HookInput) -> Vec<PathBuf> {
    let mut paths = Vec::new();
    for key in ["file_path", "path", "notebook_path"] {
        if let Some(path) = input.tool_input.get(key).and_then(|value| value.as_str()) {
            paths.push(resolve_hook_path(root, path));
        }
    }
    paths
}

fn resolve_hook_path(root: &Path, value: &str) -> PathBuf {
    let path = PathBuf::from(value);
    if path.is_absolute() {
        path
    } else {
        root.join(path)
    }
}

fn normalize_path_for_policy(root: &Path, path: &Path) -> String {
    let relative = path.strip_prefix(root).unwrap_or(path);
    trim_dot(&relative.to_string_lossy().replace('\\', "/"))
}

fn trim_dot(value: &str) -> String {
    value
        .strip_prefix("./")
        .or_else(|| value.strip_prefix(".\\"))
        .unwrap_or(value)
        .to_string()
}

fn path_matches_any(path: &str, patterns: &[String]) -> bool {
    patterns.iter().any(|pattern| path_matches(path, pattern))
}

fn path_matches(path: &str, pattern: &str) -> bool {
    let pattern = trim_dot(&pattern.replace('\\', "/"));
    if pattern.ends_with('/') {
        return path == pattern.trim_end_matches('/') || path.starts_with(&pattern);
    }
    if !pattern.contains('*') {
        return path == pattern;
    }

    let mut remaining = path;
    let mut parts = pattern.split('*').peekable();
    if let Some(first) = parts.next() {
        if !first.is_empty() {
            let Some(rest) = remaining.strip_prefix(first) else {
                return false;
            };
            remaining = rest;
        }
    }

    while let Some(part) = parts.next() {
        if part.is_empty() {
            continue;
        }
        let Some(index) = remaining.find(part) else {
            return false;
        };
        remaining = &remaining[index + part.len()..];
        if parts.peek().is_none() && !pattern.ends_with('*') {
            return remaining.is_empty();
        }
    }
    true
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn matches_exact_and_prefix_paths() {
        assert!(path_matches("GOAL.html", "GOAL.html"));
        assert!(path_matches("migrations/001.sql", "migrations/"));
        assert!(!path_matches("src/main.rs", "GOAL.html"));
    }
}

File: src/hooks.ts
import { relative, resolve } from "path";
import { goalIntegrity, readPolicy } from "./io";
import { DECISIONS_PATH, FAILURES_PATH, GOAL_PATH, HUMAN_LOG_PATH, PROGRESS_PATH } from "./paths";
import { HookInputSchema } from "./schemas";

export type HookDecision =
  | { decision: "approve" }
  | { decision: "block"; reason: string };

export function pretooluse(root: string, rawInput: string): HookDecision {
  if (rawInput.trim() === "") {
    return { decision: "approve" };
  }

  const input = HookInputSchema.parse(JSON.parse(rawInput));
  const toolInput =
    input.tool_input && typeof input.tool_input === "object" && !Array.isArray(input.tool_input)
      ? (input.tool_input as Record<string, unknown>)
      : {};
  const policy = readPolicy(root);
  if (!policy.active) {
    return { decision: "approve" };
  }

  if (isMutatingTool(input.tool_name)) {
    try {
      goalIntegrity(root);
    } catch (error) {
      return { decision: "block", reason: `${GOAL_PATH} changed while loop is active: ${formatError(error)}` };
    }
  }

  if (input.tool_name.toLowerCase() === "bash") {
    const command = typeof toolInput.command === "string" ? toolInput.command : undefined;
    if (command) {
      const lowerCommand = command.toLowerCase();
      for (const denied of policy.denyShellSubstrings) {
        if (lowerCommand.includes(denied.toLowerCase())) {
          return { decision: "block", reason: `command contains denied substring ${JSON.stringify(denied)}` };
        }
      }
      const appendOnlyLogPath = directAppendOnlyLogWrite(command);
      if (appendOnlyLogPath) {
        return {
          decision: "block",
          reason: `${appendOnlyLogPath} is append-only; use yoloop log append instead of direct shell writes`,
        };
      }
    }
  }

  for (const path of hookFilePaths(root, toolInput)) {
    const normalized = normalizePathForPolicy(root, path);
    if (pathMatchesAny(normalized, appendOnlyLogPaths())) {
      return { decision: "block", reason: `${normalized} is append-only; use yoloop log append` };
    }
    if (pathMatchesAny(normalized, policy.immutablePaths)) {
      return { decision: "block", reason: `${normalized} is immutable while yoloop is active` };
    }
    if (pathMatchesAny(normalized, policy.protectedPathsWhileActive)) {
      return { decision: "block", reason: `${normalized} is protected while yoloop is active` };
    }
  }

  return { decision: "approve" };
}

export function printHookDecision(decision: HookDecision): void {
  console.log(JSON.stringify(decision));
}

function isMutatingTool(toolName: string): boolean {
  return ["bash", "write", "edit", "multiedit", "notebookedit"].includes(toolName.toLowerCase());
}

function hookFilePaths(root: string, toolInput: Record<string, unknown>): string[] {
  const paths: string[] = [];
  for (const key of ["file_path", "path", "notebook_path"]) {
    const value = toolInput[key];
    if (typeof value === "string") {
      paths.push(resolve(root, value));
    }
  }
  return paths;
}

function normalizePathForPolicy(root: string, path: string): string {
  return trimDot(relative(root, path).replace(/\\/g, "/"));
}

function trimDot(value: string): string {
  return value.replace(/^[.][\\/]/, "");
}

function pathMatchesAny(path: string, patterns: string[]): boolean {
  return patterns.some((pattern) => pathMatches(path, pattern));
}

function directAppendOnlyLogWrite(command: string): string | undefined {
  if (!looksLikeShellWrite(command)) {
    return undefined;
  }
  const lower = command.toLowerCase().replace(/\\/g, "/");
  return appendOnlyLogPaths().find((path) => lower.includes(path.toLowerCase()));
}

function looksLikeShellWrite(command: string): boolean {
  const lower = command.toLowerCase();
  return [">", ">>", "set-content", "add-content", "out-file", "tee "].some((token) => lower.includes(token));
}

function appendOnlyLogPaths(): string[] {
  return [HUMAN_LOG_PATH, PROGRESS_PATH, FAILURES_PATH, DECISIONS_PATH].map((path) => path.replace(/\\/g, "/"));
}

function pathMatches(path: string, pattern: string): boolean {
  const normalizedPattern = trimDot(pattern.replace(/\\/g, "/"));
  if (normalizedPattern.endsWith("/")) {
    return path === normalizedPattern.slice(0, -1) || path.startsWith(normalizedPattern);
  }
  if (!normalizedPattern.includes("*")) {
    return path === normalizedPattern;
  }
  const regex = new RegExp(`^${escapeRegex(normalizedPattern).replace(/\\\*/g, ".*")}$`);
  return regex.test(path);
}

function escapeRegex(value: string): string {
  return value.replace(/[.+?^${}()|[\]\\]/g, "\\$&");
}

function formatError(error: unknown): string {
  return error instanceof Error ? error.message : String(error);
}
